docs/setup/options.md
Configuration Options
=====================

Visit the [Configuration File](configure.md) section for information about
how to use these configuration options.

Workspace
---------

> This is the location where DTrack stores data. This includes saved recordings,
> audio clips, trained models, temporary data, etc.
>
> !!! option "Default Value: `"_workspace/"`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | string  | workspace              | DTRACK\_WORKSPACE         |


Workspace Keep Temp
-------------------

> Keep temporary data after extracted for review.
>
> !!! option "Default Value: `false`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | boolean | keep\_temp             | DTRACK\_KEEP\_TEMP        |

Record Audio Device
-------------------

> The device identifier used to record audio.
>
> !!! option "Default Value: `"plughw"`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | string  | audio\_device          | RECORD\_AUDIO\_DEVICE     |
>
> - **List audio devices:** `ffmpeg -loglevel warning -sources alsa`
>     + Sample Output:
>
>             Auto-detected sources for alsa:
>             null [Discard all samples (playback) or generate zero samples (capture)]
>             hw:CARD=Snowflake,DEV=0 [Direct hardware device without any conversions]
>             plughw:CARD=Snowflake,DEV=0 [Hardware device with all software conversions]
>             default:CARD=Snowflake [Default Audio Device]
>             sysdefault:CARD=Snowflake [Default Audio Device]
>             front:CARD=Snowflake,DEV=0 [Front output / input]
>             dsnoop:CARD=Snowflake,DEV=0 [Direct sample snooping device]
>
>     + Look for devices with: "<u>software conversions</u>"
>     + It is likely worth taking time to test available options.

Record Audio Options
--------------------

> Additional ffmpeg options used for audio capture.
>
> !!! option "Default Value: `[ "-f", "alsa" ]`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | list    | audio\_options         | RECORD\_AUDIO\_OPTIONS    |

Record Video Device
-------------------

> The device identifier used to record video.
>
> !!! option "Default Value: `"/dev/video0"`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | string  | video\_device          | RECORD\_VIDEO\_DEVICE     |
>
> - **List video devices:** `ls -1 /dev/video*`
>     + Sample Output:
>
>             /dev/video0
>             /dev/video1
>             /dev/video2
>             /dev/video3
>
> - **List capabilities:** `ffmpeg -hide_banner -f v4l2 -list_formats all -i <DEVICE>`
>     + Sample Output:
>
>             [v4l2 @ 0x5612] Compressed:       mjpeg: Motion-JPEG : 640x360 640x480 960x540 1024x576 1280x720 1920x1080 2560x1440 3840x2160
>             [v4l2 @ 0x5612] Compressed:        h264:       H.264 : 640x360 640x480 960x540 1024x576 1280x720 1920x1080 2560x1440 3840x2160
>             [v4l2 @ 0x5612] Compressed: Unsupported:        HEVC : 640x360 640x480 960x540 1024x576 1280x720 1920x1080 2560x1440 3840x2160
>             [v4l2 @ 0x5612] Raw       :     yuyv422:  YUYV 4:2:2 : 640x360 640x480 960x540 1024x576 1280x720 1920x1080
>             [in#0 @ 0x5612] Error opening input: Immediate exit requested
>             Error opening input file /dev/video2.
>
>     + This output shows a camera capable of recording 4K video.

Record Video Options
--------------------

> Additional ffmpeg options used for video capture.
>
> List video devices using: `ls /dev/video*`
>
> List video device capabilities using: `v4l2-ctl -d $DEVICE --list-formats-ext`;
> look for `'H264' (H.264, compressed)`.
>
> !!! option "Default Value: `[ "-f", "v4l2", "-input_format", "h264", "-video_size", "1920x1080", "-framerate", "20" ]`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | list    | video\_options         | RECORD\_VIDEO\_OPTIONS    |

Record Video Timestamp
----------------------

> Long string that defines the datestamp that is added on top of recorded video.
>
> The default value uses `Monospace Bold`, set to a bright `Red, 48-Point` font.
> This font is provided by the `fonts-freefont-ttf` package,
>
> This filter is applied to every frame, so it is possible to add `hflip,vflip,`
> to the start of the string, which will result in an image that is flipped both
> horizontally and vertically--perfect for an upside-down camera.
>
> !!! option "Default Value: `"drawtext=fontfile=/usr/share/fonts/truetype/freefont/FreeMonoBold.ttf:text=%{localtime}:fontcolor=red@0.9:x=7:y=7:fontsize=48"`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | string  | video\_timestamp       | RECORD\_VIDEO\_TIMESTAMP  |

Record Video Advanced
---------------------

> Video creation arguments that ultimately control the final recorded video quality.
>
> A sensible range for 1080p video is between `3M` and `8M`, with `3.5M` being the
> observed upper limit on a Raspberry Pi 5--due to lack of hardware **en**coder.
>
> !!! option "Default Value: `[ "libx264", "-crf", "23", "-preset", "fast", "-tune", "zerolatency", "-maxrate", "3M", "-bufsize", "24M" ]`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | list    | video\_advanced        | RECORD\_VIDEO\_ADVANCED   |

Record Inspect Models
---------------------

> List of tags used for training, and trained models used for automatic detection.
>
> !!! option "Default Value: `[ ]`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | list    | inspect\_models        | RECORD\_INSPECT\_MODELS   |

Record Inspect Backlog
----------------------

> Defines the maximum number of "segments" that can back up in each queue
> before newly recorded audio is discarded from queue.
>
> !!! option "Default Value: `5`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | integer | inspect\_backlog       | RECORD\_INSPECT\_BACKLOG  |

Record Inspect Trust
----------------------

> Confidence level required to assume a match was found.
>
> !!! option "Default Value: `0.50`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | decimal | inspect\_trust         | RECORD\_INSPECT\_TRUST    |

Record Duration
---------------

> Length of time to use for each recording.
>
> - Very short recordings will suffer inadequate compression and are harder to review.
> - Very long recordings will be much larger and more difficult to submit.
>
> !!! option "Default Value: `"00:10:00"`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | string  | record\_duration       | RECORD\_DURATION          |

Train Batch Size
----------------

> Number of samples to use in each round of training.
>
> !!! option "Default Value: `16`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | integer | train\_batch\_size     | TRAIN\_BATCH\_SIZE        |

Train Epochs
------------

> Maximum number of learning attempts before stopping.
>
> !!! option "Default Value: `200`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | integer | train\_epochs          | TRAIN\_EPOCHS             |

Train Patience
--------------

> Maximum number of attempts to build a better model before stopping early.
>
> !!! option "Default Value: `10`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | integer | train\_patience        | TRAIN\_PATIENCE           |

Train Rate
----------

> Step size used to adjust learning. This may be decreased during training
> to prevent over-fitting data samples.
>
> !!! option "Default Value: `0.0001`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | decimal | train\_rate            | TRAIN\_RATE               |


src/state/config.go
// Application Configuration
//
// ALL VARIABLES:
// - Define Runtime variables and types:	Application_Configuration
// - Default value of Runtime variables:	Load_Configuration
// - Map JSON Settings to Runtime variables:	Application_Configuration
// - Map Environment vars to Runtime variables:	Environment_Configation_Map
// - Help text showing JSON/EnvVar/Default:	Show_Help
package state

import (
	// DTrack
	"dtrack/log"

	// Standard
	"encoding/json"
	"os"
	"reflect"
	"strconv"
)

// Master object holding loaded configuration data
var Runtime Application_Configuration

// Map json configuration to Runtime
// Defaults set in load_config()
type Application_Configuration struct {
	Workspace              string   `json:"workspace"`
	Workspace_Keep_Temp    bool     `json:"keep_temp"`
	Record_Audio_Device    string   `json:"audio_device"`
	Record_Audio_Options   []string `json:"audio_options"`
	Record_Video_Device    string   `json:"video_device"`
	Record_Video_Options   []string `json:"video_options"`
	Record_Video_Timestamp string   `json:"video_timestamp"`
	Record_Video_Advanced  []string `json:"video_advanced"`
	Record_Inspect_Models  []string `json:"inspect_models"`
	Has_Models             bool
	Record_Inspect_Backlog int      `json:"inspect_backlog"`
	Record_Inspect_Trust   float64  `json:"inspect_trust"`
	Record_Duration        string   `json:"record_duration"`
	Train_Batch_Size       int      `json:"train_batch_size"`
	Train_Epochs           int      `json:"train_epochs"`
	Train_Patience         int      `json:"train_patience"`
	Train_Rate             float64  `json:"train_rate"`
}

// Map environment variables to Runtime
var Environment_Configation_Map = map[string]string{
	"DTRACK_WORKSPACE":       "Workspace",
	"DTRACK_KEEP_TEMP":       "Workspace_Keep_Temp",
	"RECORD_AUDIO_DEVICE":    "Record_Audio_Device",
	"RECORD_AUDIO_OPTIONS":   "Record_Audio_Options",
	"RECORD_VIDEO_DEVICE":    "Record_Video_Device",
	"RECORD_VIDEO_OPTIONS":   "Record_Video_Options",
	"RECORD_VIDEO_ADVANCED":  "Record_Video_Advanced",
	"RECORD_INSPECT_MODELS":  "Record_Inspect_Models",
	"RECORD_INSPECT_BACKLOG": "Record_Inspect_Backlog",
	"RECORD_INSPECT_TRUST":   "Record_Inspect_Trust",
	"RECORD_DURATION":        "Record_Duration",
	"TRAIN_BATCH_SIZE":       "Train_Batch_Size",
	"TRAIN_EPOCHS":           "Train_Epochs",
	"TRAIN_PATIENCE":         "Train_Patience",
	"TRAIN_RATE":             "Train_Rate",
}

// Loads Runtime configuration data into current state
func Load_Configuration(config_path string) {
	// Default configuration values
	cfg := Application_Configuration{
		Workspace:              "_workspace",
		Workspace_Keep_Temp:    false,
		Record_Audio_Device:    "plughw",
		Record_Audio_Options:   []string{"-f", "alsa"},
		Record_Video_Device:    "/dev/video0",
		Record_Video_Options:   []string{
			"-f", "v4l2", "-input_format", "h264",
			"-video_size", "1920x1080", "-framerate", "20"},
		Record_Video_Timestamp: "drawtext=fontfile=/usr/share/fonts/truetype/freefont/" +
			"FreeMonoBold.ttf:text=%{localtime}:fontcolor=red@0.9:x=7:y=7:fontsize=48",
		Record_Video_Advanced: []string{
			"libx264", "-crf", "23", "-preset", "fast", "-tune", "zerolatency",
			"-maxrate", "3M", "-bufsize", "24M"},
		Record_Duration:        "00:10:00",
		Record_Inspect_Models:  []string{},
		Record_Inspect_Backlog: 5,
		Record_Inspect_Trust:   0.50,
		Train_Batch_Size:       16,
		Train_Epochs:           200,
		Train_Patience:         10,
		Train_Rate:             0.0001,
	}

	// Check for configuration file
	log.Debug("Loading configuration values from: %s", config_path)
	if _, err := os.Stat(config_path); err != nil {
		log.Info("Configuration file not found; using defaults.")
		Runtime = cfg
		return
	}

	// Load configuration file
	file_data, err := os.ReadFile(config_path)
	if err != nil {
		log.Die("Error opening configuration file; ABORT!")
	}

	// Merge configuration values into cfg
	if err := json.Unmarshal(file_data, &cfg); err != nil {
		log.Die("Failed to parse configuration as JSON; ABORT!")
	}

	// Search for known environment variables
	for env_key, conf_field := range Environment_Configation_Map {
		if env_value := os.Getenv(env_key); env_value != "" {
			log.Debug("Environment variable found: %s", env_key)
			field := reflect.ValueOf(&cfg).Elem().FieldByName(conf_field)
			if !field.IsValid() || !field.CanSet() {
				log.Die("Invalid field: %s", conf_field)
			}

			// Merge environment variables into cfg
			switch field.Kind() {
			case reflect.String:
				field.SetString(env_value)
			case reflect.Int:
				if intVal, err := strconv.Atoi(env_value); err == nil {
					field.SetInt(int64(intVal))
				} else {
					log.Die("%s is not Integer", env_key)
				}
			case reflect.Float64:
				if intVal, err := strconv.Atoi(env_value); err == nil {
					field.SetFloat(float64(intVal))
				} else {
					log.Die("%s is not Float64", env_key)
				}
			case reflect.Bool:
				if boolVal, err := strconv.ParseBool(env_value); err == nil {
					field.SetBool(boolVal)
				} else {
					log.Die("%s is not Boolean", env_key)
				}
			default:
				log.Die("Unexpected field type for %s", conf_field)
			}
		}
	}

	// Helper variables
	cfg.Has_Models = len(cfg.Record_Inspect_Models) > 0

	// Update session variables
	Runtime = cfg
}
